File: backend/src/services/salary_generation.py
"""Pure salary generation logic — no DynamoDB dependency."""

import uuid
import calendar
from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Tuple


def generate_salaries_for_month(
    training_events: List[Dict[str, Any]],
    coaches: List[Dict[str, Any]],
    existing_salaries: List[Dict[str, Any]],
    month: str,
    academy: str,
) -> Tuple[List[Dict[str, Any]], int]:
    """Generate salary records from training events for a given month.

    Args:
        training_events: schedule entries of type 'training' for the target month.
            Each must have 'id', 'coachId', 'teamId', 'date'.
        coaches: list of coach dicts. Each must have 'id', 'name', 'salaryPerTraining'.
        existing_salaries: salary records already created for this month.
            Each must have 'scheduleEventId' and 'coachId'.
        month: target month in ``YYYY-MM`` format.
        academy: academy identifier to stamp on new records.

    Returns:
        A tuple ``(new_salaries, skipped_count)``.
    """
    # Build a lookup of coaches by id
    coaches_by_id: Dict[str, Dict[str, Any]] = {c['id']: c for c in coaches}

    # Build a set of (scheduleEventId, coachId) already covered
    existing_keys = {
        (s['scheduleEventId'], s['coachId'])
        for s in existing_salaries
    }

    # Compute last day of month for default dueDate
    year, mon = int(month[:4]), int(month[5:7])
    last_day = calendar.monthrange(year, mon)[1]
    default_due_date = f"{month}-{last_day:02d}"

    new_salaries: List[Dict[str, Any]] = []
    skipped = 0
    now = datetime.utcnow().isoformat()

    for event in training_events:
        event_id = event['id']
        coach_id = event.get('coachId', '')

        if (event_id, coach_id) in existing_keys:
            skipped += 1
            continue

        coach = coaches_by_id.get(coach_id, {})
        salary_rate = Decimal(str(coach.get('salaryPerTraining', 0)))

        # Resolve team name from the event or fall back
        team_name = event.get('teamName', '')

        salary = {
            'id': str(uuid.uuid4()),
            'coachId': coach_id,
            'coachName': coach.get('name', ''),
            'teamId': event.get('teamId', ''),
            'teamName': team_name,
            'scheduleEventId': event_id,
            'trainingDate': event.get('date', ''),
            'month': month,
            'amount': salary_rate,
            'originalAmount': salary_rate,
            'status': 'pending',
            'dueDate': default_due_date,
            'paidAmount': Decimal('0'),
            'academy': academy,
            'createdAt': now,
            'updatedAt': now,
        }
        new_salaries.append(salary)

    return new_salaries, skipped

File: frontend/jogafacil/src/utils/salaryGeneration.js
/**
 * Pure salary generation logic — mirrors backend/src/services/salary_generation.py
 * for frontend testability.
 */

/**
 * Returns the last day of a month given a YYYY-MM string.
 *
 * @param {string} month - format YYYY-MM
 * @returns {number} last day of the month
 */
function getLastDayOfMonth(month) {
  const [year, mon] = month.split('-').map(Number)
  return new Date(year, mon, 0).getDate()
}

/**
 * Generates salary records from training events for a given month.
 *
 * @param {Array<{ id: string, coachId: string, teamId?: string, teamName?: string, date: string }>} trainingEvents
 * @param {Array<{ id: string, name: string, salaryPerTraining?: number }>} coaches
 * @param {Array<{ scheduleEventId: string, coachId: string }>} existingSalaries
 * @param {string} month - target month in YYYY-MM format
 * @returns {{ newSalaries: Array, skippedCount: number }}
 */
export function generateSalaryRecords(trainingEvents, coaches, existingSalaries, month) {
  // Build coach lookup by id
  const coachesById = {}
  for (const coach of coaches) {
    coachesById[coach.id] = coach
  }

  // Build set of existing (scheduleEventId, coachId) keys
  const existingKeys = new Set(
    existingSalaries.map(s => `${s.scheduleEventId}|${s.coachId}`)
  )

  // Compute default due date (last day of month)
  const lastDay = getLastDayOfMonth(month)
  const defaultDueDate = `${month}-${String(lastDay).padStart(2, '0')}`

  const now = new Date().toISOString()
  const newSalaries = []
  let skippedCount = 0

  for (const event of trainingEvents) {
    const eventId = event.id
    const coachId = event.coachId || ''
    const key = `${eventId}|${coachId}`

    if (existingKeys.has(key)) {
      skippedCount++
      continue
    }

    const coach = coachesById[coachId] || {}
    const salaryRate = Number(coach.salaryPerTraining) || 0

    newSalaries.push({
      id: crypto.randomUUID(),
      coachId,
      coachName: coach.name || '',
      teamId: event.teamId || '',
      teamName: event.teamName || '',
      scheduleEventId: eventId,
      trainingDate: event.date || '',
      month,
      amount: salaryRate,
      originalAmount: salaryRate,
      status: 'pending',
      dueDate: defaultDueDate,
      paidAmount: 0,
      createdAt: now,
      updatedAt: now,
    })
  }

  return { newSalaries, skippedCount }
}
